src/eval.rs

```diff
@@ -6,46 +6,86 @@ use crate::EvalError;
 use gc::{Finalize, Gc, GcCell, Trace};
 use rnix::TextRange;
 use std::borrow::Borrow;
+use std::collections::HashMap;
 
-type ExprResult = Result<Box<Expr>, EvalError>;
+// Expressions like BinOp have the only copy of their Expr children,
+// so they use ExprResultBox. Expressions like Map, which may have
+// contents copied in multiple places, need ExprResultGc.
+type ExprResultBox = Result<Box<Expr>, EvalError>;
+type ExprResultGc = Result<Gc<Expr>, EvalError>;
 
 /// Used to lazily calculate the value of a Expr. This should be
 /// tolerant of parsing and evaluation errors from child Exprs.
-#[derive(Debug, Clone, Trace, Finalize)]
+///
+/// We store everything that we want the user to inspect. For example,
+/// the source for an attribute key-value pair includes the key so the
+/// user can hover inside dynamic keys in code like
+/// `{ "${toString (1+1)}" = 2; }`.
+#[derive(Debug, Trace, Finalize)]
 pub enum ExprSource {
+    // We want to share child MapAttrs between the ExprSource
+    // and the value map, so we use Gc.
+    Map {
+        inherits: Vec<ExprResultGc>,
+        definitions: Vec<ExprResultGc>,
```

Comment:
I don't think I understood the purpose of `inherits`/`definitions` here since we still keep a map with everything inside.

Reply:
For syntax like `{ foo.abc = 1; foo.xyz = 2; }`, the NixValue hashmap has `foo` merged, but we still want the separate syntax trees for `climb_expr`, thus the duplication. I now realize that dividing between inherit vs normal attributes is unnecessary, so I'll push code to store them together in the ExprSource

Reply:
Seems resolved now.